02_bubble/willshen8/bubble_test.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"bytes"
+	"testing"
+)
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+
+	expected := "[1 2 3 5]"
+	actual := buf.String()
+
+	if expected != actual {
+		t.Errorf("Unexpected output in main(), expected = %v, actual = %v", expected, actual)
+	}
+}
+
+func TestBubbleSort(t *testing.T) {
+	tests := []struct {
+		name     string
+		input    []int
+		expected []int
+	}{
+		{name: "Empty slice", input: []int{}, expected: []int{}},
+		{name: "Reverse sorted slice", input: []int{10, 9, 8, 7, 6, 5, 4, 3, 2, 1},
+			expected: []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
+		{name: "Already sorted slice", input: []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
+			expected: []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
+		{name: "Slice with negative number", input: []int{1, -1}, expected: []int{-1, 1}},
+	}
+
+	for _, test := range tests {
```

Comment:
Have a look at "Introducing sub tests" in [this blog](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)

Reply:
I find subtests are sometimes warranted and other times not.
Good: Different test cases (edge cases, different input types etc)
Bad: Randomized tests
In this case it would be a good idea to use them, but I will leave this in the 'do next time' bucket. since you do go through the effort to report the name in the test logs.
To run a subtest
```
t.Run(name, func(t *testing.T) {
  // sub test, written just like any other test, have access to test data
})
```